src/coreclr/jit/valuenum.cpp

```diff
@@ -11061,11 +11064,26 @@ static bool GetStaticFieldSeqAndAddress(ValueNumStore* vnStore, GenTree* tree, s
         if (fldSeq != nullptr)
         {
             assert(fldSeq->GetKind() == FieldSeq::FieldKind::SimpleStaticKnownAddress);
-            *pFseq      = fldSeq;
-            *byteOffset = vnStore->CoercedConstantValue<ssize_t>(treeVN) - fldSeq->GetOffset() + val;
-            return true;
+
+            // Is the offset within the field? There might be a sequence of adds on top of a constant static field
+            // address due to shared constant CSE such that `treeVN` is not the static field base address.
+            // Check if the entire access doesn't span across the field and any other field. This simple check is
+            // just that the access starts within the field.
+            CORINFO_FIELD_HANDLE fieldHandle = fldSeq->GetFieldHandle();
+            CORINFO_CLASS_HANDLE fieldClsHnd;
+            var_types            fieldType = eeGetFieldType(fieldHandle, &fieldClsHnd);
+            ssize_t              fieldSize = (fieldType == TYP_STRUCT) ? info.compCompHnd->getClassSize(fieldClsHnd)
+                                                          : (ssize_t)genTypeSize(fieldType);
+            ssize_t tmpByteOffset = vnStore->CoercedConstantValue<ssize_t>(treeVN) - fldSeq->GetOffset() + val;
+            if ((0 <= tmpByteOffset) && (tmpByteOffset < fieldSize))
+            {
+                *pFseq      = fldSeq;
+                *byteOffset = tmpByteOffset;
+                return true;
+            }
```

Comment:
This is the real fix. The rest is just clean-ups.

Reply:
I presume this the problem that is reproducable via Shared Constant + JitOptRepeat

I still don't think I understand the problem and it feels like this code is trying to mask some other problem or it's just some invalid IL that is doing something semi-legal, like reading out of bounds of a field (in that case `getStaticFieldContent` should just return false, if it currently hits asserts - we should probably just relax those, its contract is to return false on any non-foldable access)

The idea of this code was to simply return a field handle (as part of FieldSeq) + exact address of the content we need.

then, without validations, we ask VM via `getStaticFieldContent` to get us that content if it's valid and is immutable.

Typically, we work with trees like this:

```
IND(fieldAddr + offset) where fieldAddr has a field sequence (known static)
```
if the fieldAddr is split into pieces via shared constants, I presume the total offset should still be the same.

Maybe shared constants are doing something illegal around where exactly to leave a field sequence?

cc @jakobbotsch @SingleAccretion in case if this rings any bell

Reply:
Is a dump of where this failure occurs available?

The way the code is written (before this change), it effectively expects `getStaticFieldContent` to validate against nonsensical offsets that can occur in UB-like code.

If this occurs in not-UB-like code, that would be a bug somewhere else, not here.

Reply:
There is no UB code here.

In one test case (Linux/arm32, 
```
System.Console:<get_IsInputRedirected>g__EnsureInitialized|34_0():System.Runtime.CompilerServices.StrongBox`1[ubyte] (MethodHash=e981a3d5)
```
with OptRepeat enabled (using all the fixes from https://github.com/dotnet/runtime/pull/94250), we get an assert:
```
Assert failure(PID 31115 [0x0000798b], Thread: 31115 [0x798b]): valueOffset == 0
    File: /home/brucefo/gh/runtime2/src/coreclr/vm/jitinterface.cpp:11783
```
https://github.com/dotnet/runtime/blob/8c68710084b861c44b6868c0bd972ce24aed55dd/src/coreclr/vm/jitinterface.cpp#L11780-L11791

We start with Value Numbering:
```
***** BB01, STMT00004(before)
N003 (  6, 11) [000012] ----G-------                        *  RETURN    ref
N002 (  5, 10) [000011] n---G-------                        \--*  IND       ref
N001 (  2,  8) [000010] H-----------                           \--*  CNS_INT(h) int    -0x4D1FE0A0 static Fseq[_isStdInRedirected]

N001 [000010]   CNS_INT(h) -0x4D1FE0A0 static Fseq[_isStdInRedirected] => $c4 {Hnd const: 0x0xb2e01f60 GTF_ICON_STATIC_HDL}
    VNForHandle(_isStdInRedirected) is $c5, fieldType is ref, size = 4
      AX1: select([$82]store($2c1, $c5, $101), $c5) ==> $101.
    VNForMapSelect($2c1, $c5):ref returns $101 {JitNew($c2, $141)}
N002 [000011]   IND       => <l:$101 {JitNew($c2, $141)}, c:$142 {MemOpaque:NotInLoop}>
N003 [000012]   RETURN    => $VN.Void
```

Because of shared constant CSE, we replace the `[000010]` tree by defining a CSE base address and use an offset, as follows:
```
N006 (  6, 11)              [000022] DA--G-------                        *  STORE_LCL_VAR ref    V02 tmp2         d:1 $VN.Void                                                                                                        N005 (  6, 11)              [000014] nA--G-------                        \--*  IND       ref    <l:$100, c:$140>                                                                                                                      N004 (  3,  9)              [000037] -A----------                           \--*  COMMA     int    $c0                                                                                                                                N002 (  2,  8) CSE #01 (def)[000035] DA----------                              +--*  STORE_LCL_VAR int    V04 cse0         d:1 $VN.Void                                                                                               N001 (  2,  8)              [000013] H-----------                              |  \--*  CNS_INT(h) int    -0x4D1FE0D0 static Fseq[STDIN_FILENO] $c0                                                                                   N003 (  1,  1)              [000036] ------------                              \--*  LCL_VAR   int    V04 cse0         u:1 $c0           
...
N005 (  5,  3)              [000012] ----G-------                        *  RETURN    ref    $VN.Void                                                                                                                                 N004 (  4,  2)              [000011] n---G-------                        \--*  IND       ref    <l:$101, c:$142>
N003 (  3,  4)              [000043] -------N----                           \--*  ADD       int    $c4
N001 (  1,  1)              [000041] ------------                              +--*  LCL_VAR   int    V04 cse0         u:1 $c0
N002 (  1,  2)              [000042] ------------                              \--*  CNS_INT   int    48           
```

In the 2nd iteration of JitOptRepeat, we re-run Value Numbering on:
```
***** BB01, STMT00004(before)
N005 (  5,  3) [000012] ----G-------                        *  RETURN    ref
N004 (  4,  2) [000011] n---G-------                        \--*  IND       ref
N003 (  3,  4) [000043] -------N----                           \--*  ADD       int
N001 (  1,  1) [000041] ------------                              +--*  LCL_VAR   int    V04 cse0         u:1 (last use)
N002 (  1,  2) [000042] ------------                              \--*  CNS_INT   int    48

N001 [000041]   LCL_VAR   V04 cse0         u:1 (last use) => $c0 {Hnd const: 0x0xb2e01f30 GTF_ICON_STATIC_HDL}
N002 [000042]   CNS_INT   48 => $46 {IntCns 48}
N003 [000043]   ADD       => $47 {IntCns 0xB2E01F60}                      
```
then hit the assert.

This is because `fgGetStaticFieldSeqAndAddress` walks past the `ADD` to find the FieldSeq for the `cse0` node, and sets the byteOffset to 48, from the `CNS_INT` node. This is, of course, non-sensical for the `cse0` field. `fgValueNumberConstLoad` then passes this on to `getStaticFieldContent` which asserts because the field type is REF and we're giving a non-zero offset.

The fix here adds some additional checking -- validation checking, if you will -- that the offset is valid and within the field.

An alternative might be to converts the asserts in `getStaticFieldContent` to `if` statements: https://github.com/dotnet/runtime/pull/100320

Reply:
Here's the whole dump before the crash. [fail.txt](https://github.com/dotnet/runtime/files/14766077/fail.txt)

Reply:
Thank you - after looking at the dump I agree that this (or checking on the EE side) is the right fix. This is a case where the Jit itself creates an out-of-bounds field sequence access.

(Of course, for CQ purposes, it would be better if it didn't, but that's a different problem.)